compiler: untangle SIMD alignment assumptions

workingjubilee opened the pull request:
There were a number of puzzling assumptions being made about SIMD types and their layout that I have corrected in this diff. These are mostly no-op edits in actual fact, but they do subtly alter a pair of checks in our invariant-checking and union layout computation that rested on those peculiar assumptions. Those unfortunately stand in the way of any further actual fixes. I submit this for review, even though it's not clearly motivated without its followups, because it should still be possible to independently conclude whether this is correct.

<!-- homu-ignore:start -->
One of these checks is yours, I believe?
r? RalfJung
<!-- homu-ignore:end -->

workingjubilee committed:
compiler: Align::max_for_offset -> Align::max_aligned_factor

No functional changes.

workingjubilee committed:
compiler: `BackendRepr::inherent_{size,align} -> scalar_{size,align}`

This pair of fn was introduced to perform invariant checks for scalars.
Their current behavior doesn't mesh as well with checking SIMD types,
so change the name of the fn to reflect their actual use-case and
refactor the corresponding checks.

Also simplify the returns from Option<AbiAndPrefAlign> to Option<Align>,
because every site was mapping away the "preferred" alignment anyways.

workingjubilee committed:
rust-analyzer: use new function in (incorrect) layout computation

This is only to fix the build.

RalfJung commented on compiler/rustc_ty_utils/src/layout/invariant.rs:
```diff
@@ -66,12 +66,12 @@ pub(super) fn layout_sanity_check<'tcx>(cx: &LayoutCx<'tcx>, layout: &TyAndLayou
             assert_matches!(
                 layout.layout.backend_repr(),
-                BackendRepr::Uninhabited | BackendRepr::Memory { .. },
+                BackendRepr::Uninhabited | BackendRepr::Memory { .. } | BackendRepr::Vector { .. },
```
I have no recollection of why these two variants are permitted here. But adding `Vector` seems sus as that one is much more like the "scalar" reprs than the others. So why does this make sense?

RalfJung replied:
Seems like this came out of a refactor in https://github.com/rust-lang/rust/commit/3b1e535f36ac4c47dc91d0e3394dca72fb86db0c.

workingjubilee replied:
@RalfJung Because currently we use repr annotations to specify details relevant to the alignment, and the vast majority of alignments we can pick for these types are forgeries: only a few vectors actually have a given alignment for an architecture.

RalfJung commented:
Sorry, I don't have any knowledge of the vast majority of the code being touched here. I won't be able to review this. I'm not even sure who our experts here are, maybe @bjorn3 or @nikic ?

r? compiler

RalfJung commented on compiler/rustc_abi/src/lib.rs:
```diff
@@ -1463,16 +1467,18 @@ impl BackendRepr {
-        Some(match *self {
-            BackendRepr::Scalar(s) => s.align(cx),
-            BackendRepr::ScalarPair(s1, s2) => s1.align(cx).max(s2.align(cx)),
+    /// fixed alignment of this repr, if any is mandated
```
This seems to say the same thing as the old sentence, just without being a properly capitalized sentence?

RalfJung replied:
And the trailing dot as well. ;)

RalfJung commented on compiler/rustc_abi/src/lib.rs:
```diff
@@ -1485,12 +1491,12 @@ impl BackendRepr {
                 // May have some padding between the pair.
                 let field2_offset = s1.size(cx).align_to(s2.align(cx).abi);
-                (field2_offset + s2.size(cx)).align_to(self.inherent_align(cx)?.abi)
+                (field2_offset + s2.size(cx)).align_to(self.required_align(cx)?)
```
Something seems *really wrong* if we can't compute the offset of the 2nd field of a ScalarPair. I would have expected an `unwrap`. But I guess this is pre-existing so :shrug: I don't understand any of this I guess.^^

I originally wrote the layout sanity check, but not because I understand layouts particularly well -- rather, because I don't understand them at all and wanted to see some structure.

workingjubilee commented on compiler/rustc_abi/src/lib.rs:
```diff
@@ -1463,16 +1467,18 @@ impl BackendRepr {
     }
 
-    /// Returns the fixed alignment of this ABI, if any is mandated.
-    pub fn inherent_align<C: HasDataLayout>(&self, cx: &C) -> Option<AbiAndPrefAlign> {
```
I don't think "inherent size" is a good name either for what it's actually returning.

workingjubilee replied:
Previously it returned a fuzzy notion of "inherent alignment" that included many numbers that were "we made up a number because it seemed good" rather than having a hard relationship with requirements.

Now it only should return alignments if they are actually required by the psABI.

RalfJung replied:
I think the history here is roughly as follows:
There are various parts in the compiler that rely on `BackendRepr::Scalar{,Pair}` types having *exactly* the size and alignment that would also be returned by the corresponding functions on [`Scalar`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_abi/enum.Scalar.html). We had some bugs where those did not actually match. So I added a sanity check to ensure they do match. "inherent size/align" might not be the best term for that, but "required align" is worse IMO as one could argue that if the type's alignment is *larger* than the "required align", the requirement is satisfied -- but we need full equality.

Then at some point I wondered about `BackendRepr::Vector`. I have absolutely no clue what the structure is there, so I threw together some random assertions until everything passed. Clearly I didn't know what I was doing. :)

So I think the main question we have to figure out is whether any part of the compiler relies on re-computing the size/alignment of `BackendRepr::Vector` types from the `element` and `count`. If not, we can restrict this sanity check to `Scalar`/`ScalarPair`.

RalfJung replied:
Is "natural alignment" wrong here? Not sure what exactly that means.

> Maybe just scalar_alignment would be fine...

Seems fine to me.

workingjubilee replied:
So, "natural alignment" has two possible meanings that I have seen used:
- "whatever the alignment is by default for this type"
- "an alignment equal to the size of the type"

And because Murphy's Law seems to really like applying itself wherever ABI comes up, the meaning you think it means is almost invariably not the meaning being used by the person or document you are interacting with when it comes up.

workingjubilee replied:
Anyway that seems like a reasonable plan overall!

RalfJung commented:
Looking at https://github.com/rust-lang/rust/issues/130402, `repr(align)` was also implicated in the issues around `Scalar`/`ScalarPair` that I mentioned above. For those types, the fix is to use `BackendRepr::Aggregate` instead when `repr(align)` is present. Maybe that is also the right fix for SIMD types? Or maybe those can have their alignment overridden in `Layout` without causing havoc in the rest of the compiler?

RalfJung commented:
(Unrelated, maybe we should rename `BackendRepr::Vector` to `BackendRepr::SimdVector`, since for most Rustaceans "vector" will refer to `Vec`.)

workingjubilee commented:
> Looking at https://github.com/rust-lang/rust/issues/130402, repr(align) was also implicated in the issues around Scalar/ScalarPair that I mentioned above. For those types, the fix is to use BackendRepr::Aggregate instead when repr(align) is present. Maybe that is also the right fix for SIMD types? Or maybe those can have their alignment overridden in Layout without causing havoc in the rest of the compiler?

What exactly is "causing havoc in the rest of the compiler"? The code that I am adjusting is the layout code.

workingjubilee commented:
I also didn't see you mention any particular issues?

RalfJung commented:
Yeah sorry I was referring to the issues mentioned [here](https://github.com/rust-lang/rust/pull/137256#discussion_r1961245821).

workingjubilee commented:
will revise accordingly

@rustbot author

workingjubilee committed:
compiler: split vector_align into cabi and llvmlike

Introduce a pair of functions that actually describe what they do,
because it wasn't clear the alignment is sometimes a forgery.

workingjubilee commented:
@rustbot ready

workingjubilee commented:
Will conflict with https://github.com/rust-lang/rust/pull/136985 so should not be put in the queue yet but is otherwise ready for review

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13424204250/job/37503734486) [(plain)](https://github.com/rust-lang/rust/commit/a96ec2a0674475a3cf88f0e9d631b370da2eb4c3/checks/37503734486/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04
ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.859 Building wheels for collected packages: reuse
#12 2.860   Building wheel for reuse (pyproject.toml): started
#12 3.073   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.074   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=0c2fd2aaf7b0bf8d6e131220aff14712a774c2ca462f3204d25460cbcf610b63
#12 3.075   Stored in directory: /tmp/pip-ephem-wheel-cache-i1dtl_tc/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.077 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.468 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.468 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.010 Collecting virtualenv
#12 4.010 Collecting virtualenv
#12 4.051   Downloading virtualenv-20.29.2-py3-none-any.whl (4.3 MB)
#12 4.156      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 41.9 MB/s eta 0:00:00
#12 4.214 Collecting filelock<4,>=3.12.2
#12 4.219   Downloading filelock-3.17.0-py3-none-any.whl (16 kB)
#12 4.238 Collecting distlib<1,>=0.3.7
#12 4.241   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.250      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 74.2 MB/s eta 0:00:00
#12 4.282 Collecting platformdirs<5,>=3.9.1
#12 4.287   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#12 4.369 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.553 Successfully installed distlib-0.3.9 filelock-3.17.0 platformdirs-4.3.6 virtualenv-20.29.2
#12 DONE 4.6s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      133056 kB
DirectMap2M:     6158336 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-02-08/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13424343887/job/37504152082) [(plain)](https://github.com/rust-lang/rust/commit/efb89c975d40f5177174a20b5b608b09fc491503/checks/37504152082/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 27.5s done
#21 DONE 40.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
failures:

---- [ui] tests/ui/layout/debug.rs stdout ----

error: Error: expected failure status (Some(1)) but received status Some(101).
status: exit status: 101
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/layout/debug.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/layout/debug" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
--- stderr -------------------------------
error: unions cannot have zero fields
##[error]  --> /checkout/tests/ui/layout/debug.rs:83:1
   |
   |
LL | union EmptyUnion {} //~ ERROR: has an unknown layout
   | ^^^^^^^^^^^^^^^^^^^

error: layout_of(E) = Layout {
           size: Size(12 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: Memory {
               sized: true,
           },
           fields: Arbitrary {
---
               tag_encoding: Direct,
               tag_field: 0,
               variants: [
                   Layout {
                       size: Size(4 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(1 bytes),
                           pref: Align(8 bytes),
                       abi: Memory {
                           sized: true,
                       },
                       fields: Arbitrary {
---
                       unadjusted_abi_align: Align(1 bytes),
                       randomization_seed: 5882656968055370760,
                   },
                   Layout {
                       size: Size(12 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(4 bytes),
                           pref: Align(8 bytes),
                       abi: Uninhabited,
                       fields: Arbitrary {
                           offsets: [
                               Size(4 bytes),
---
           randomization_seed: 17647407962802756638,
       }
  --> /checkout/tests/ui/layout/debug.rs:8:1
   |
LL | enum E { Foo, Bar(!, i32, i32) } //~ ERROR: layout_of

error: layout_of(S) = Layout {
           size: Size(8 bytes),
           align: AbiAndPrefAlign {
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: ScalarPair(
               Initialized {
                   value: Int(
                       I32,
---
           randomization_seed: 7269008155437992722,
       }
  --> /checkout/tests/ui/layout/debug.rs:11:1
   |
LL | struct S { f1: i32, f2: (), f3: i32 } //~ ERROR: layout_of


error: layout_of(U) = Layout {
           size: Size(8 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: Memory {
               sized: true,
           },
           fields: Union(
---
           randomization_seed: 4825121522533703520,
       }
  --> /checkout/tests/ui/layout/debug.rs:14:1
   |
LL | union U { f1: (i32, i32), f3: i32 } //~ ERROR: layout_of


error: layout_of(Result<i32, i32>) = Layout {
           size: Size(8 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(8 bytes),
           abi: ScalarPair(
               Initialized {
                   value: Int(
                       I32,
---
               tag_encoding: Direct,
               tag_field: 0,
               variants: [
                   Layout {
                       size: Size(8 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(4 bytes),
                           pref: Align(8 bytes),
                       abi: ScalarPair(
                           Initialized {
                               value: Int(
                                   I32,
---
                       unadjusted_abi_align: Align(4 bytes),
                       randomization_seed: 11796406305399895283,
                   },
                   Layout {
                       size: Size(8 bytes),
                       align: AbiAndPrefAlign {
                           abi: Align(4 bytes),
                           pref: Align(8 bytes),
                       abi: ScalarPair(
                           Initialized {
                               value: Int(
                                   I32,
---
           randomization_seed: 16942756313171812054,
       }
  --> /checkout/tests/ui/layout/debug.rs:17:1
   |
LL | type Test = Result<i32, i32>; //~ ERROR: layout_of


error: layout_of(i32) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(4 bytes),
               pref: Align(4 bytes),
           abi: Scalar(
               Initialized {
                   value: Int(
                       I32,
---
           randomization_seed: 18446462603027873795,
       }
  --> /checkout/tests/ui/layout/debug.rs:20:1
   |
LL | type T = impl std::fmt::Debug; //~ ERROR: layout_of

error: layout_of(V) = Layout {
           size: Size(2 bytes),
           align: AbiAndPrefAlign {
           align: AbiAndPrefAlign {
               abi: Align(2 bytes),
               pref: Align(8 bytes),
           abi: Memory {
               sized: true,
           },
           fields: Union(
---
           randomization_seed: 309364857495531842,
       }
  --> /checkout/tests/ui/layout/debug.rs:26:1
   |
LL | pub union V { //~ ERROR: layout_of


error: layout_of(W) = Layout {
           size: Size(2 bytes),
           align: AbiAndPrefAlign {
               abi: Align(2 bytes),
               pref: Align(8 bytes),
           abi: Memory {
               sized: true,
           },
           fields: Union(
---
           randomization_seed: 7914191794798900433,
       }
  --> /checkout/tests/ui/layout/debug.rs:32:1
   |
LL | pub union W { //~ ERROR: layout_of


error: layout_of(Y) = Layout {
           size: Size(0 bytes),
           align: AbiAndPrefAlign {
               abi: Align(2 bytes),
               pref: Align(8 bytes),
           abi: Memory {
               sized: true,
           },
           fields: Union(
---
           randomization_seed: 15087507716613604296,
       }
  --> /checkout/tests/ui/layout/debug.rs:38:1
   |
LL | pub union Y { //~ ERROR: layout_of


error: layout_of(P1) = Layout {
           size: Size(4 bytes),
           align: AbiAndPrefAlign {
               abi: Align(1 bytes),
               pref: Align(1 bytes),
           abi: Memory {
               sized: true,
           },
           fields: Union(
---
           randomization_seed: 3973676115800971815,
       }
  --> /checkout/tests/ui/layout/debug.rs:45:1
   |
LL | union P1 { x: u32 } //~ ERROR: layout_of


error: layout_of(P2) = Layout {
           size: Size(8 bytes),
           align: AbiAndPrefAlign {
               abi: Align(1 bytes),
               pref: Align(1 bytes),
           abi: Memory {
               sized: true,
           },
           fields: Union(
---
           randomization_seed: 13035393740343907081,
       }
  --> /checkout/tests/ui/layout/debug.rs:49:1
   |
LL | union P2 { x: (u32, u32) } //~ ERROR: layout_of


thread 'rustc' panicked at compiler/rustc_ty_utils/src/layout/invariant.rs:239:17:
assertion failed: align >= element_align
assertion failed: align >= element_align
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: rustc_ty_utils::layout::invariant::layout_sanity_check
   4: rustc_ty_utils::layout::layout_of
      [... omitted 3 frames ...]
   5: rustc_middle::query::plumbing::query_get_at::<rustc_query_system::query::caches::DefaultCache<rustc_middle::ty::PseudoCanonicalInput<rustc_middle::ty::Ty>, rustc_middle::query::erase::Erased<[u8; 16]>>>
   7: <rustc_session::session::Session>::time::<(), rustc_interface::passes::run_required_analyses::{closure#5}>
   8: rustc_interface::passes::analysis
   8: rustc_interface::passes::analysis
      [... omitted 3 frames ...]
   9: <std::thread::local::LocalKey<core::cell::Cell<*const ()>>>::with::<rustc_middle::ty::context::tls::enter_context<<rustc_middle::ty::context::GlobalCtxt>::enter<rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>::{closure#1}, core::option::Option<rustc_interface::queries::Linker>>::{closure#0}, core::option::Option<rustc_interface::queries::Linker>>
  10: <rustc_middle::ty::context::TyCtxt>::create_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}>
  11: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  12: <alloc::boxed::Box<dyn for<'a> core::ops::function::FnOnce<(&'a rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, &'a std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena<'a>>, &'a rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena<'a>>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}), Output = core::option::Option<rustc_interface::queries::Linker>>> as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once
  14: rustc_span::create_session_globals_then::<(), rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}::{closure#0}>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: the compiler unexpectedly panicked. this is a bug.
error: the compiler unexpectedly panicked. this is a bug.

note: using internal features is not supported and expected to cause internal compiler errors when used incorrectly

note: rustc 1.87.0-nightly (a36067584 2025-02-19) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=/cargo -Z ignore-directory-in-diagnostics-source-blocks=/checkout/vendor -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0
query stack during panic:
#0 [layout_of] computing layout of `P3`
#1 [analysis] running analysis passes on this crate
end of query stack
```

</details>

workingjubilee commented:
I suppose then

r? @bjorn3

RalfJung commented on compiler/rustc_ty_utils/src/layout/invariant.rs:
```diff
@@ -65,31 +65,30 @@ pub(super) fn layout_sanity_check<'tcx>(cx: &LayoutCx<'tcx>, layout: &TyAndLayou
         // Verify per-ABI invariants
         match layout.layout.backend_repr() {
             BackendRepr::Scalar(_) => {
+                // we already asserted this
```
```suggestion
                // These must always be present for `Scalar` types.
```

workingjubilee replied:
Applied

RalfJung commented on compiler/rustc_ty_utils/src/layout/invariant.rs:
```diff
@@ -231,9 +230,15 @@ pub(super) fn layout_sanity_check<'tcx>(cx: &LayoutCx<'tcx>, layout: &TyAndLayou
+                let element_size = element.size(cx);
+                // Currently, vectors must always be aligned to at least their elements:
+                assert!(align >= element_align);
+                // And the size has to be element * count, of course
+                assert!(size >= (element_size * count).align_to(align));
```
Why is this not `==`?

workingjubilee replied:
...y'know after the `.align_to` it should in fact be `==`

RalfJung reviewed (commented):
Modulo one question, r=me on the `compiler/rustc_ty_utils/src/layout/invariant.rs` part. I can't really speak for the rest.

bjorn3 commented on compiler/rustc_abi/src/lib.rs:
```diff
@@ -408,16 +408,21 @@ impl TargetDataLayout {
-        // That is, use the size, rounded up to a power of 2.
-        AbiAndPrefAlign::new(Align::from_bytes(vec_size.bytes().next_power_of_two()).unwrap())
+    pub fn cabi_vector_align(&self, vec_size: Size) -> Option<AbiAndPrefAlign> {
+        self.vector_align
```
Why is this public when it is called only in `llvmlike_vector_align`?

bjorn3 commented:
r=me,RalfJung with both our questions answered.

bjorn3 commented:
@bors r=bjorn3,RalfJung

bors commented:
:pushpin: Commit b6117492b28767c189ce060fff9dfaa99160f4ad has been approved by `bjorn3,RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3,RalfJung b6117492b28767c189ce060fff9dfaa99160f4ad -->
<!-- homu: {"type":"Approved","sha":"b6117492b28767c189ce060fff9dfaa99160f4ad","approver":"bjorn3,RalfJung","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13441362757/job/37556365636) [(plain)](https://github.com/rust-lang/rust/commit/b6117492b28767c189ce060fff9dfaa99160f4ad/checks/37556365636/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04
ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 3.125 Building wheels for collected packages: reuse
#12 3.126   Building wheel for reuse (pyproject.toml): started
#12 3.350   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.351   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=0c2fd2aaf7b0bf8d6e131220aff14712a774c2ca462f3204d25460cbcf610b63
#12 3.351   Stored in directory: /tmp/pip-ephem-wheel-cache-k8yuni9c/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.354 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.777 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.777 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.356 Collecting virtualenv
#12 4.356 Collecting virtualenv
#12 4.428   Downloading virtualenv-20.29.2-py3-none-any.whl (4.3 MB)
#12 4.533      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 42.9 MB/s eta 0:00:00
#12 4.596 Collecting filelock<4,>=3.12.2
#12 4.611   Downloading filelock-3.17.0-py3-none-any.whl (16 kB)
#12 4.648 Collecting platformdirs<5,>=3.9.1
#12 4.663   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#12 4.683 Collecting distlib<1,>=0.3.7
#12 4.706      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 86.4 MB/s eta 0:00:00
#12 4.706      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 86.4 MB/s eta 0:00:00
#12 4.793 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.982 Successfully installed distlib-0.3.9 filelock-3.17.0 platformdirs-4.3.6 virtualenv-20.29.2
#12 DONE 5.1s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      126912 kB
DirectMap2M:     8261632 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-02-08/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
```

</details>

RalfJung commented:
@bors r-

workingjubilee commented:
HOW

workingjubilee commented:
@bors r=bjorn3,RalfJung

bors commented:
:pushpin: Commit f964368500e88340909cd21c2a72ef009bccfd3b has been approved by `bjorn3,RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3,RalfJung f964368500e88340909cd21c2a72ef009bccfd3b -->
<!-- homu: {"type":"Approved","sha":"f964368500e88340909cd21c2a72ef009bccfd3b","approver":"bjorn3,RalfJung","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
huh, unexpected. :melting_face: 

oh well, rebased on https://github.com/rust-lang/rust/pull/136985 since it will break this anyways

workingjubilee commented:
@bors r=bjorn3,RalfJung

bors commented:
:pushpin: Commit 46f7a7d378972b4034118f798657a6d35091a4ca has been approved by `bjorn3,RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3,RalfJung 46f7a7d378972b4034118f798657a6d35091a4ca -->
<!-- homu: {"type":"Approved","sha":"46f7a7d378972b4034118f798657a6d35091a4ca","approver":"bjorn3,RalfJung","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
